PhysicsTools/PatAlgos/python/tools/jetTools.py

```diff
@@ -597,9 +597,17 @@ def setupBTagging(process, jetSource, pfCandidates, explicitJTA, pvSource, svSou
                                     btag.pixelClusterTagInfos.clone(jets = jetSource, vertices=pvSource),
                                     process, task)
 
-            if 'Puppi' in jetSource.value() and pfCandidates.value() == 'particleFlow' and\
-	       ('pfBoostedDouble' in btagInfo or 'SecondaryVertex' in btagInfo):
-                _btagInfo = getattr(process, btagPrefix+btagInfo+labelName+postfix)
+            if 'pfBoostedDouble' in btagInfo or 'SecondaryVertex' in btagInfo:
+              _btagInfo = getattr(process, btagPrefix+btagInfo+labelName+postfix)
+              if pfCandidates.value() == 'packedPFCandidates':
+                _btagInfo.weights = cms.InputTag("packedpuppi")
+                if not hasattr(process,"packedpuppi"):
+                  from CommonTools.PileupAlgos.Puppi_cff import puppi
+                  addToProcessAndTask('packedpuppi', puppi.clone(
+                        useExistingWeights = True,
+                        candName = 'packedPFCandidates',
+                        vertexName = cms.InputTag('offlineSlimmedPrimaryVertices')) , process, task)
+              else:
                 _btagInfo.weights = cms.InputTag("puppi")
```

Comment:
please drop type specifications for all parameters which already exist.
This is a safer way to protect from parameter name mistakes and will also help in possible parameter migrations.

```_btagInfo.weights = "packedpuppi"```
```vertexName = "offlineSlimmedPrimaryVertices"```
```_btagInfo.weights = "puppi"```